Removed Debug and Releases. Refactored the entire code base.

StudioLE opened the pull request:
@WaynePatrickDalton , @2adicted and @CaoimheLoftus 

I'm currently trying to get my head around how C# works. I don't think the Releases and Debug files need to be in the repository. These should be compiled on Build, correct?

Therefore I've removed them with commit: https://github.com/teamtreedyn/Binoculars/pull/1/commits/99526307a9ec5b5cbbf10aa1057fb8f7b52c188d

StudioLE commented:
I then went through all the code and did the following:

- Added comments throughout the code to make it easier for new contributors to understand.
- Added @todo items to be worked on.
- Changed how and when ExportData is stored.
- As Environment variables like IP and computer name don't change there's no point in setting them every run. Instead they're now set during start up.
- Removed WIP code.

Take a look at the changes when you get a chance and let me know if it should be merged into master.

brencass commented:
Two things that might be worthwhile looking at/changing.

1. If you want to use a 24 hour clock, it should be"HH" instead of "hh".  Would it be worth indicating all times "local time" for the system and linked to country 

More details on what the caps/non caps variations can mean: https://docs.microsoft.com/en-us/dotnet/api/system.datetime.tostring?view=netframework-4.7.2.

2. If you are looking at how to possibly utilise a "settings" file within your extension. I implemented it as follows within my DynaStandard extension. It means that it will be adaptive to if the extension DLL/settings file location could be in any possible locations(Packages, ViewExtension, server?). Though they do need to be in the same folder.

![image](https://user-images.githubusercontent.com/14090637/55690465-11c51b80-598a-11e9-8986-2062cf8bc8d1.png)

radumg commented on TrackerEvents.cs:
```diff
@@ -71,16 +42,22 @@ public static void UnregisterEventHandlers()
+            ExportData.filename = obj.Name;
+
+            // Display a MessageBox to the user
+            // @todo Remove this as it's for testing purposes only
```
instead of removing the `MessageBox`, use a [C# IF conditional compilation statement](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives/preprocessor-if). This way, when compiliing with `Debug` build setup, you see it, and when you compile with `Release` you don't.

StudioLE replied:
Thanks for this @radumg it's exactly what we need. I am however running into an issue with it. Perhaps you can advise? See issue: https://github.com/teamtreedyn/Binoculars/issues/22

StudioLE replied:
Marking this resolved as #if DEBUG added in https://github.com/teamtreedyn/Binoculars/commit/9046734bf8ef5ace6a5dc3a943674d0975750ec1

radumg commented on TrackerEvents.cs:
```diff
@@ -71,16 +42,22 @@ public static void UnregisterEventHandlers()
         {
+            // Subscribe our event handler methods to Dynamo
             model.RunCompleted += OnGraphRun;
             model.EvaluationCompleted += OnEvaulationCompleted;
```
```suggestion
            model.EvaluationCompleted += OnEvaluationCompleted;
```

radumg commented on TrackerEvents.cs:
```diff
@@ -91,32 +68,22 @@ internal static void RegisterRunEventHandlers(DynamoModel model)
         /// <param name="e"></param>
         private static void OnEvaulationCompleted(object sender, EvaluationCompletedEventArgs e)
```
```suggestion
        private static void OnEvaluationCompleted(object sender, EvaluationCompletedEventArgs e)
```

radumg commented on TrackerEvents.cs:
```diff
@@ -91,32 +68,22 @@ internal static void RegisterRunEventHandlers(DynamoModel model)
+            // When graph evaluation is complete
+            // Get all the data we want to store and then store it
+            var dataToExport = ExportData.Export();
+            ExportSheets.Execute(dataToExport);
```
Both these feel very un-C# like, would suggest something that is more OO and not use so many `static` methods.

radumg commented on TrackerEvents.cs:
```diff
@@ -91,32 +68,22 @@ internal static void RegisterRunEventHandlers(DynamoModel model)
-
-                var dataToExport = ExportData.Export(FileName, DynamoVersion);
-                ExportSheets.Execute(dataToExport);
             }
```
why does this not work ? I'd either
- trace the issue & fix
- leave it as is
- delete the entire contents of the methods, leaving the `if` in there doesn't serve any purpose

StudioLE replied:
Because of my commits I think the line reference above has changed and should refer to:

https://github.com/teamtreedyn/Binoculars/blob/master/Events.cs#L37
```c#
        private static void OnGraphRun(object sender, bool success)
        {
            // DOES NOT WORK
            if (success)
            {
            }
        }
```

The answers to this is Dynamo's `RunCompleted` event is simply never called by Dynamo:  https://github.com/DynamoDS/Dynamo/issues/9602

Therefore, we will remove this event and continue to use `EvaluationCompleted`

StudioLE replied:
Fixed in https://github.com/teamtreedyn/Binoculars/commit/5df70272967bfc013f6b71a0cd51f19e9165f0b0 therefore marking resolved.

radumg commented on TrackerExtenstion.cs:
```diff
@@ -38,14 +39,48 @@ public void Startup(StartupParams sp)
+            // @todo Subscribe to an event after the Dynamo home screen is shown
+
+            // Display a MessageBox to the user
+            // @todo Use a view framework to improve the UI/UX
```
or don't :) the current box does the trick nicely and has no overhead - unless the interaction needs to be more complex, with checkboxes, etc, i wouldn't add anything just for a message box.

StudioLE replied:
Discussion continues as issue https://github.com/teamtreedyn/Binoculars/issues/19 therefore marking this resolved.

radumg commented on TrackerExtenstion.cs:
```diff
@@ -38,14 +39,48 @@ public void Startup(StartupParams sp)
             MessageBox.Show(message, title, MessageBoxButton.OK, MessageBoxImage.Exclamation);
+
+            // @todo if user does not consent, don't store..
+            if (true) {
```
instead of wrapping a whole bunch of code in an `if` statement, bail early.
```suggestion
            if (!true) return;
```

StudioLE replied:
Very good point! I primarily work with javascript lately so I'd completely forgotten to take advantage  of blocking/synchronous methods

StudioLE replied:
Closing in favour of issue https://github.com/teamtreedyn/Binoculars/issues/20

radumg commented on TrackerExtenstion.cs:
```diff
@@ -38,14 +39,48 @@ public void Startup(StartupParams sp)
             // we can register our own events that will be triggered when specific things happen in Dynamo
             // a reference to the ReadyParams is needed to do this, so we pass it on
-            TrackerEvents.RegisterEventHandlers(rp);            
+            TrackerEvents.RegisterEventHandlers(rp);
```
if this stays here, the `if` that bails early would mean this never gets triggered - which i assume is the desired behaviour given someone hasn't opted in.

StudioLE replied:
Closing in favour of issue https://github.com/teamtreedyn/Binoculars/issues/20

radumg commented on TrackerViewExtension.cs:
```diff
@@ -69,6 +69,8 @@ public void TrackerMenuItems()
             sayHelloMenuItem.Click += (sender, args) =>
             {
+                // Display a MessageBox to the user
+                // @todo Use a view framework to improve the UI/UX
```
how would it improve UX ? All view frameworks come with baggage, i'd steer clear for such a simple thing.

StudioLE replied:
Discussion continues as issue https://github.com/teamtreedyn/Binoculars/issues/19 therefore marking this resolved.

radumg commented on TrackerExtenstion.cs:
```diff
@@ -38,14 +39,48 @@ public void Startup(StartupParams sp)
+                ExportData.computerName = Environment.UserName;
+                
+                // @todo Define the revitVersion, leave blank or null if opened from any other environment (Sandbox, Civil3D etc)
+                ExportData.revitVersion = "2019.0.2";
```
use the Dynamo host information for this.

StudioLE replied:
New issue created: https://github.com/teamtreedyn/Binoculars/issues/21 therefore I'm marking resolved.

radumg commented on TrackerExtenstion.cs:
```diff
@@ -38,14 +39,48 @@ public void Startup(StartupParams sp)
+                // Request the IP and geolocation from ipinfo.io
+                WebClient client = new WebClient();
+                client.Headers.Set("Accept", "application/json");
+                var json = client.DownloadString("https://ipinfo.io");
```
wouldn't put this in the `Ready` event, since it'd be better if it was deferred to be executed when needed - which is not on startup, but on event collection or export

StudioLE replied:
I'm currently measuring the response time of ipinfo.io as between 180 and 300ms on a decent connection which is quite significant, hence why I'd rather it was completed during Dynamo startup rather than on EvaluationCompleted. The presumption is that anyone loading Dynamo will at some point run a graph so deferring it makes little difference but slowing down their first run will be noticed and should be minimised.

radumg commented on ExportData.cs:
```diff
@@ -20,39 +20,27 @@ namespace Tracker
-            computerName = Environment.UserName;
+            // Set the date
+            // @todo "hh" is incorrectly returning values in the afternoon. It should be "14" not "2" etc for any hour after midday
             date = DateTime.Now.ToString("yyyy-MM-dd hh:mm:ss");
```
I;'d suggest to use the [ISO8601 standard format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings#the-round-trip-o-o-format-specifier) instead of custom formats.

StudioLE replied:
New issue created: https://github.com/teamtreedyn/Binoculars/issues/17 Therefore I'm marking this resolved.

radumg commented on ExportData.cs:
```diff
@@ -20,39 +20,27 @@ namespace Tracker
         private static string date;
 
-        internal static IList<IList<object>> Export(string filename, string dynamoversion)
+        internal static IList<IList<object>> Export()
```
the purpose of this method seems to be : `create` an object that has all the data needed for export.
To make this more idiomatic C#, I would suggest to make a class `public class ExportData` with all the fields enumerated above (`user`, `computer`, `name`, `ip`, `etc`). The constructor of the class would initialise these fields with the correct values from the `Environment` class or by fetching IP etc.

StudioLE replied:
Agreed, this has mostly been resolved with the creation of the Data.Collect() method https://github.com/teamtreedyn/Binoculars/blob/master/Data.cs#L35

Although I opted for a static class. 

Discussion on this can continue as issue https://github.com/teamtreedyn/Binoculars/issues/18 therefore I'm marking this resolved.

radumg commented on ExportData.cs:
```diff
@@ -20,39 +20,27 @@ namespace Tracker
+            // @todo "hh" is incorrectly returning values in the afternoon. It should be "14" not "2" etc for any hour after midday
             date = DateTime.Now.ToString("yyyy-MM-dd hh:mm:ss");
 
-            WebClient client = new WebClient();
```
Web requests can be slow & behave unexpectedly. Also, unlike `user name` and other fields that could change between events, this is not likely to change, so it's a prime candidate to be cached (such as storing it in a `static` prop).

StudioLE replied:
Agreed! Hopefully this is resolved with the new static Data class.  Continue discussion in issue https://github.com/teamtreedyn/Binoculars/issues/18 Therefore marking as resolved.

radumg commented on ExportData.cs:
```diff
@@ -91,48 +79,24 @@ public static void Execute(IList<IList<object>> list)
             });
 
             // Define request parameters.
+            // @todo Fetch these from a config.json file or environment variables?
```
create a class called `BinocularsConfig` that represents your settings (for example, it would have a field called `GoogleSpreadsheetId`). This can be stored in `json` files and easily read & `de/serialised` in the Extension constructor.

StudioLE replied:
Agreed, discussion continues as issue https://github.com/teamtreedyn/Binoculars/issues/12 therefore marking resolved.

radumg commented on ExportData.cs:
```diff
@@ -91,48 +79,24 @@ public static void Execute(IList<IList<object>> list)
             var rqst = service.Spreadsheets.Values.Append(vRange, spreadsheetId, rng);
             rqst.ValueInputOption = SpreadsheetsResource.ValuesResource.AppendRequest.ValueInputOptionEnum.USERENTERED;
-
             rqst.Execute();
```
wrap in `try {} catch {}`

StudioLE replied:
Agreed. I've created an issue: https://github.com/teamtreedyn/Binoculars/issues/23 therefore marking as resolved.

2adicted commented:
Hey Radu,

Thank you for the thorough investigation, that's so much information there! Much appreciated man!

Just a question - can you point me to a repo or documentation regarding the Dynamo Host .. object? How to initialize or get that to begin to harvest information? Thanks again!

Deyan

radumg commented:
@2adicted check out Host-related stuff here : 
https://github.com/DynamoDS/Dynamo/blob/2ef67e17e77bd12067bd33cda90c9de7ef38097b/src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs#L365

You have access to that on the `DynamoViewModel` which you get on the view extension side ;)

brencass commented:
@radumg may correct me on this or part of it.

I think you need to have the HostName into a check(eg host name == "Dynamo Revit")  that will the initialise the revit checker once required. Once you get to the revit section of your extension you will need to start looking at revit api and referencing the correct dlls into your solution.

I have quickly just ran the host name and version in a extension i have and i get the following(with code as well below). Not it was blank when it was ran from sandbox or from visual studio. but had the below when ran in revit 2019.

Dynamo From Revit:
![Untitled](https://user-images.githubusercontent.com/14090637/55820777-d21a4300-5af3-11e9-9051-d271aaf107b9.png)

Dynamo Sandbox(ran from visual studio):
![Capture2](https://user-images.githubusercontent.com/14090637/55820869-0988ef80-5af4-11e9-95fb-08a1e72730bf.PNG)

Code Used:
![Capture](https://user-images.githubusercontent.com/14090637/55820787-d6def700-5af3-11e9-8867-c342bd629049.PNG)

2adicted commented:
@brencass @radumg Yes, that's what I mean. I think we need to import the Persistence namespace, the Document reference (the way we would do in Python node in Dynamo?) as a workaround. 

btw, the level of Dynamo extensions is not too far from what it was this time last year. They can definitely give it some looouve I recon.

brencass commented:
@2adicted Yes you can import the Persistence namespace and if you go that route you will need to import the relevant dll's. These will be from the dynamo revit folder(eg revitservices), and the relevant ones from the normal revit install folder(eg Revitapi).

Quick example and shows code: https://autode.sk/2D9oQrB

Do not forget to set the copying of dll's you reference to false in your solution.

2adicted commented:
@brencass Yeees, that's exactly it, brother! You've got it :D

radumg commented:
also don't forget to check if not running sandbox - those DLLs might not be available and it'll bring it crashing down 💥

StudioLE commented:
I had some time to kill on trains over the the weekend so thought I'd do a further refactor of the full codebase. Let me know your thoughts on the following and whether they are changes worth keeping.

- Removed TrackerExtension.cs and moved its logic in to TrackerViewExtension.cs as it seems extraneous to have both.
- Renamed namespace, files and methods from Tracker to Binoculars
- Added ExportData.Record() method to collect environment data (per Radu's recommendation, although I couldn't find a constructor workflow that I liked..)
- Removed OnCurrentWorkspaceChanged event as it didn't seem necessary
- Fixed a few things that were broken

I've got some additional changes in mind, including move to an MVC (Model View Controller)  pattern to simplify the extension. I'll explain more soon.

2adicted commented:
Hey Laurence. 

I merged your changes, nice refactoring there. I tried doing a similar task and I got somewhere until it broke and refused to work ever again. 

I pulled your changes, rebuild the definition, started it and it worked once ... and that's it, same thing happened - not working anymore. 

I will have a fresh look at some other time, too frustrated to deal with rubbish dynamo now.

StudioLE commented:
> I pulled your changes, rebuild the definition, started it and it worked once ... and that's it, same thing happened - not working anymore.

Try deleting both extensions from `%appdata%\Dynamo\Dynamo Core\2.0\packages` and delete everything in the bin folder where you're building the application. That was causing a ton of issues for me because the old Tracker files kept being copied across.

2adicted commented:
I have a hunch about the Google Sheets module ... not sure if you did the same thing, but I logged in with a different email than the one I was invited to collaborate with. Both times this coincided with the app not working, but I cannot be sure as it is failing silently at the moment.

We need to get each of those modules polished to shine and really know them inside out. 

Still good job on the refactoring, it doesn't take away from the fact that the code is starting to look better and better. In my version, I oped for a more object oriented class which was not static. I can show you later after we get back on track.
